### PR TITLE
feat: parallel phase 1 and 2

### PR DESCRIPTION
**Pull request by pszemraj**
## MuonClip correctness + FSDP2 single-node support + BF16 runtime stabilization

**What changed and why:**

**1. Optimizer correctness & API cleanup**
Muon parameter grouping now uses identity/module membership instead of name-prefix heuristics, restricting Muon to transformer hidden 2D weights and routing everything else to AdamW. Orthogonalization and normalization scaling are decoupled; NS epsilon is `1e-7`; normalization is an explicit `norm_factor` policy (`spectral` default, plus `match_rms_adamw`, `none`, `legacy_compat` for bisection safety).

**2. FSDP2 distributed support**
Owner-compute path for DTensor `Shard(0)` via `gather`/`scatter` collectives with uneven-shard handling (row-count exchange + pad/unpad) and cached load-balanced owner assignment. QK clipping auto-disables under FSDP2 with one-time warning. Compatibility checks now allow FSDP2 while hard-rejecting FSDP1 and DeepSpeed ZeRO2.

**3. BF16 runtime stabilization (`torch 2.10.0+cu128`, RTX 5090)**
`flash_attn_varlen + bf16` crashed at startup with `CUBLAS_STATUS_INVALID_VALUE` in QKV linear. A startup probe now detects this and switches to `cublaslt`; if BF16 still fails, falls back to `mixed_precision="no"` + SDPA. Integrated into pretrain, GLUE, and contrastive trainers.

**Validation**
- `206 passed, 8 skipped, 18 subtests passed` (CPU deterministic)
- Requested 600s single-GPU run completes cleanly to `exit 124` (timeout), no crash
- Previous `CUBLAS_STATUS_INVALID_VALUE` path no longer reproduced

**Remaining:** Manual multi-GPU FSDP2 golden test (`torchrun --nproc_per_node=2`) pending in colleague's environment to confirm distributed parity and resume behavior.

### CODE REVIEW

**Comment by pszemraj** [comment]
Revised guide:

```sh

# 0. Sanity check: this host must actually expose 2 GPUs
python -c "import torch; print(torch.cuda.device_count())"

# 1. Optional correctness precheck (manual golden)
torchrun --standalone --nproc_per_node=2 \
  tests/manual/test_muonclip_fsdp2_golden.py
```

```sh
# 1b. Optional production resume-path smoke
torchrun --standalone --nproc_per_node=2 \
  tests/manual/test_muonclip_accelerate_fsdp2_resume.py
```


```sh
# 2. MuonClip config (original file)
CUDA_VISIBLE_DEVICES=0,1 \
accelerate launch \
  --num_processes 2 --num_machines 1 \
  --mixed_precision bf16 \
  --dynamo_backend no \
  --use_fsdp --fsdp_version 2 \
  --fsdp_auto_wrap_policy TRANSFORMER_BASED_WRAP \
  --fsdp_transformer_layer_cls_to_wrap EncoderBlock \
  scripts/pretraining/pretrain.py \
  configs/pretraining/pretrain_neobert100m_smollm2data_muonclip.yaml \
  --wandb.name "neobert-muon-parallel-testing"
```

Use this for distributed

```sh
# 3. Explicit no-clipping variant
CUDA_VISIBLE_DEVICES=0,1 \
accelerate launch \
  --num_processes 2 --num_machines 1 \
  --mixed_precision bf16 \
  --dynamo_backend no \
  --use_fsdp --fsdp_version 2 \
  --fsdp_auto_wrap_policy TRANSFORMER_BASED_WRAP \
  --fsdp_transformer_layer_cls_to_wrap EncoderBlock \
  scripts/pretraining/pretrain.py \
  configs/pretraining/pretrain_neobert100m_smollm2data_muonclip_noclip.yaml \
  --wandb.name "neobert-muon-parallel-testing"
```

```sh
# 4. 600s crash smoke
CUDA_VISIBLE_DEVICES=0,1 timeout 600s \
accelerate launch \
  --num_processes 2 --num_machines 1 \
  --mixed_precision bf16 \
  --dynamo_backend no \
  --use_fsdp --fsdp_version 2 \
  --fsdp_auto_wrap_policy TRANSFORMER_BASED_WRAP \
  --fsdp_transformer_layer_cls_to_wrap EncoderBlock \
  scripts/pretraining/pretrain.py \
  configs/pretraining/pretrain_neobert100m_smollm2data_muonclip_noclip.yaml \
  --wandb.name "neobert-muon-parallel-testing"
```

**Review by amazingvince — APPROVED** [review]
